File: MovieServer/MovieServer.API/MovieServer.Shared/src/Dispatcher/Commands/CommandDispatcher.ts
import { ICommand, ICommandHandler } from "@Shared/Commands";
import ICommandDispatcher from "../Commands/ICommandDispatcher";
import { Injectable, InversifyContainer } from "@Shared/IoC";
import { Container } from "inversify";
import { GetSymbol } from "@Shared/IoC/Utils";

@Injectable
export default class CommandDispatcher implements ICommandDispatcher {

    private readonly _container: Container;
    constructor(
        @InversifyContainer container: Container
    ) {
        this._container = container;

        this.DispatchAsync = this.DispatchAsync.bind(this);
    }
    DispatchAsync<TCommand extends ICommand>(command: TCommand): Promise<any> {
        // get command handler based on TCommand
        console.log(`Commander: ICommandHandler<${command.constructor.name}>`);
        const symbol = GetSymbol(`ICommandHandler<${command.constructor.name}>`)
        const handler = this._container.get<ICommandHandler<TCommand>>(symbol as symbol);
        // handle command
        return handler.HandleAsync(command);
    }

}

File: OrderServer/OrderServer.API/OrderServer.Shared/Dispatcher/Commands/CommandDispatcher.cs
using Microsoft.Extensions.DependencyInjection;
using OrderServer.Shared.Commands;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace OrderServer.Shared.Dispatcher.Commands
{
    public class CommandDispatcher : ICommandDispatcher
    {
        private readonly IServiceProvider _serviceProvider;
        public CommandDispatcher(
            IServiceProvider serviceProvider)
        {
            _serviceProvider = serviceProvider;
        }
        public Task DispatchAsync<TCommand>(TCommand command) where TCommand : ICommand
        {
            // get service 
            using var scope = _serviceProvider.CreateScope();
            var handler = scope.ServiceProvider.GetService<ICommandHandler<TCommand>>()!;
            // handle async
            return handler.HandleAsync(command);
        }
    }
}
